### PR TITLE
Protect form a large number of instances started simultaneously, Fix #14

### LINKED ISSUE #14: Add provisions to protect form a large number of instances started simultaneously
This might be in the form of touching a temporary/hidden file and making sure there is a grace period between processes sending data

### PR DESCRIPTION
**Pull request by minborg**

### CODE REVIEW

File: src/main/java/net/openhft/chronicle/analytics/internal/AbstractGoogleAnalytics.java [hunk 1]
```diff
@@ -25,16 +25,20 @@
...
 
+import static net.openhft.chronicle.analytics.internal.FilesUtil.isNowSameAsLastUsedFileTimeStampSecond;
+
 abstract class AbstractGoogleAnalytics implements Analytics {
 
     private final AnalyticsConfiguration configuration;
     private final String clientId;
     private final AtomicLong lastSendAttemptNs = new AtomicLong();
     private final AtomicInteger sentMessages = new AtomicInteger();
+    private final boolean createdSameAsLastUsedDileTimeStampSecond;
```
Comment (alamar) [inline src/main/java/net/openhft/chronicle/analytics/internal/AbstractGoogleAnalytics.java]: "FileTimeStamp"

File: src/main/java/net/openhft/chronicle/analytics/internal/AbstractGoogleAnalytics.java [hunk 2]
```diff
@@ -53,6 +57,10 @@ public void sendEvent(@NotNull final String name, @NotNull final Map<String, Str
     abstract void httpSend(@NotNull String eventName, @NotNull final Map<String, String> eventParameters);
 
     boolean attemptToSend() {
+        if (createdSameAsLastUsedDileTimeStampSecond) {
```
Comment (alamar) [inline src/main/java/net/openhft/chronicle/analytics/internal/AbstractGoogleAnalytics.java]: I guess it should be checked here and not in constructor?

File: src/main/java/net/openhft/chronicle/analytics/internal/FilesUtil.java [hunk 3]
```diff
@@ -40,4 +42,47 @@ static String acquireClientId(@NotNull final String clientIdFileName, @NotNull f
...
+            Files.delete(lastPath());
+        } catch (IOException ignore) {
+            // Just for testing so we do not care so much
+        }
+    }
+
+    private static Path lastPath() {
+        final String fileName = Optional.ofNullable(System.getProperty("user.home"))
+                .orElse(".") +
+                "/.chronicle.analytics.last";
```
Comment (alamar) [inline src/main/java/net/openhft/chronicle/analytics/internal/FilesUtil.java]: Makes sense to move to constant

File: src/main/java/net/openhft/chronicle/analytics/internal/FilesUtil.java [hunk 4]
```diff
@@ -40,4 +42,47 @@ static String acquireClientId(@NotNull final String clientIdFileName, @NotNull f
         }
         return id;
     }
+
+    static boolean isNowSameAsLastUsedFileTimeStampSecond() {
```
Comment (alamar) [inline src/main/java/net/openhft/chronicle/analytics/internal/FilesUtil.java]: Can we have more precise granularity - makes sense to just return second of day and compare outside this method.

**Review by alamar — CHANGES_REQUESTED** [review]

File: src/main/java/net/openhft/chronicle/analytics/internal/AbstractGoogleAnalytics.java (hunk 2 above)
Comment (minborg) [reply src/main/java/net/openhft/chronicle/analytics/internal/AbstractGoogleAnalytics.java]: No. It is checked at construction time. That way, only one instance on the machine will actually send data of those created in the same second.

**Review by alamar — APPROVED** [review]

File: src/main/java/net/openhft/chronicle/analytics/internal/FilesUtil.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+    static void touchLastContent(Path path) {
+        try {
+            Files.write(path, Integer.toString(LocalTime.now().toSecondOfDay()).getBytes(UTF_8));
+        } catch (IOException ignored) {
+            // Not much we can do about it...
+        }
+    }
+
+    // Just for test purposes
```
Comment (JerryShea) [inline src/main/java/net/openhft/chronicle/analytics/internal/FilesUtil.java:75]: `@TestOnly`

**Pull request merged**